rfcs/0000-tx-valid-since/0000-template.md

```diff
@@ -0,0 +1,45 @@
+---
+Number: "0000"
+Category: Standards Track
+Status: Proposal
+Author: Jinyang Jiang
+Organization: Nervos Foundation
+Created: 2019-03-11
+---
+
+# Transaction valid since
+
+## Abstract
+
+This RFC suggests adding a new consensus rule to prevent a cell to be spent before a certain block timestamp or a block number.
+
+## Summary 
+
+Transaction adds a new `u64` type field `valid_since`, which prevent the transaction to be mined before a certain point, the highest bit of `valid_since` is `type flag`, the other 63 bits represent the actual value `V`, the consensus to validate this field described as follow:
```

Comment:
This is not necessarily a problem but just want to throw a discussion here: if we look at core, we don't (I think?) yet have a field that encode multiple entries now. One example is `reference` vs `binary` in Script, while we can keep both in a single `Vec<u8>`, we opt to put them in 2 fields and use `Option` to wrap them so we know if a value is present there.

I was thinking maybe we can do the same thing here: we can use a separate flag value and 2 more values for block number and timestamp individually. CFB should help us ensure one missing item won't need the extra space. I know this will be more repetitive than a single field here, but what we gain here, is more clarity in the code. Thoughts?

Reply:
I think 1-bit flag is simple enough and can be clearly express in code. If CFB incurs the same 1 bit cost on separate fields I'm ok with either way, otherwise I prefer 1-bit flag for succinct.

Reply:
Sounds fair to me

Reply:
Flatbuffers has a significant overhead to encode a union.

Reply:
Use `union` looks good, it's a more modern way to represent `valid_since`, `union` needs a `u8` enum type to represent different value, I think it is not a big deal.

Updated, use `union` to represent the `valid_since` field.

Reply:
If each block is 1s, `1 << 62` will take 146235604338.768 years to run out. I'm not sure why you're uncomfortable with such a long time.

And I don't think `1 << 63` is much better than `1 << 62`.

Reply:
Succint is a very important goal for messages that will be transmitted on network. The waste will accumulate quickly if we don't pay attention.

Reply:
Fair enough, I'll revert to `1-bit` proposal.